teps/0076-array-result-types.md

```diff
@@ -0,0 +1,620 @@
+  see every Task including a step with protobuf tools in it or we'd need to include some by default)
+* `xml` (seems more verbose than we'd need and afaik in cloud native tools isn't in wide use)
+* `csv` (would also be easy to use but afaik it wouldn't support more complex types like dicts? or we'd need to add some
+  kind of syntax on top of the csv format to define more complex types)
+
+This proposal suggests using `json` because:
+* It's human-readable
+* It's possible for Tasks to write json content without needing extra tools
+* We've added json specific support to other parts of our APIs (e.g. [in triggers](https://github.com/tektoncd/triggers/blob/f2380d7d33666547ee7ed965730cf7dbc1aa5421/docs/triggerbindings.md#accessing-data-in-http-json-payloads))
+* Kubernetes supports `yaml` and `json`, so it makes sense to narrow our choices to at least these, as opposed to
+  introducing a new format into the mix.
+
+We may want to consider also adding support for [JSON text sequences](#also-support-json-text-sequences) in future.
+
+### Notes/Caveats
+
+* Introducing typing to results raises questions about
+  [validating the structure of the results emitted by Tasks](#validating-json-results-at-runtime)
+* The way results are currently supported [constrains the size of results that can be emitted](#size-limits)
+
+### Risks and Mitigations
+
+* Changing the type of Task Result values from string to being ArrayOrString might be backwards incompatible
+    * Mitigation: perform a manual test to make sure this isn't backwards incompatible (reconcile a TaskRun with a string
+      result both before and after the change - e.g. create a TaskRun before the change, then restart the controller so
+      it is re-reconciled; revisit this proposal if we can't find a way to make the change and be backward compatible)
+    * Note this will be a change for folks consuming the library, e.g. the CLI, as a result implementing this TEP
+      should include working with the CLI team to pull in the updates and make sure they are not negatively impacted
+
+## Design Details
+
+* [Emitting array results](#emitting-array-results)
+* [Indexing into arrays with variable replacement](#indexing-into-arrays-with-variable-replacement)
+* [Using array results with variable replacement](#using-array-results-with-variable-replacement)
+
+### Emitting array results
+
+We add support for writing json content to `/tekton/results/resultName`, supporting strings and arrays of strings.
+
+For example, say we want the value of the Task foo's result `animals` to be "cat", "dog" and "squirrel":
+
+1. The Task would declare a result with `type:array`:
+
+  ```yaml
+  apiVersion: tekton.dev/v1beta1
+  kind: Task
+  metadata:
+    name: task-with-parameters
+  spec:
+    params: ...
+    steps: ...
+    results:
+    - name: animals
+      type: array # this field is new for results
```

Comment:
I know the whole proposal is about making array results, but this "type: array" is in conflict with "type: object" — have you considered a different attribute like "multiple: true" or something?  

With both 0075 and 0076, you'd be able to define something like:

```
  name: animals
  type: object
  properties:
    name:
      type: string
    weight:
      type: number
  multiple: true
```

and emit (using json text sequence, of course)

```
cat <<EOF > $(results.animals.path)
{ "name": "cat", weight: 33}
{ "name": "dog", weight: 44}
```

Reply:
Interesting - i was imagining the syntax being more json schema aligned to support the use case you're describing (an array of objects), e.g. something like this (based on https://json-schema.org/understanding-json-schema/reference/array.html and https://json-schema.org/learn/miscellaneous-examples.html#arrays-of-things):

```
  name: animals
  type: array
  items: {
    type: object
    properties:
      name:
        type: string
      weight:
        type: number
  }
```

And if we supported json text sequences, i would think the schema would look the same, the difference would be in how the results written by the Task were formatted? (i.e. im not seeing these features as being in conflict but let me know if I'm missing something)

However I have intentionally tried to keep nested types out of the scope of this proposal and TEP-0075 as well. I would want to make sure nothing we propose here would stop us from adding nested type support in the future so if you see this being incompatible please continue to raise it.

Personally I would still lean toward using the json schema syntax for this vs. introducing a new `multiple` field.